src/shadowbox/server/shared_metrics.ts

```diff
@@ -82,18 +94,61 @@ export class PrometheusUsageMetrics implements UsageMetrics {
 
   constructor(private prometheusClient: PrometheusClient) {}
 
+  private async queryUsage(
+    timeSeriesSelector: string,
+    deltaSecs: number
+  ): Promise<PrometheusQueryResult> {
+    const query = `
+      sum(increase(${timeSeriesSelector}[${deltaSecs}s]))
+      by (${PROMETHEUS_COUNTRY_LABEL}, ${PROMETHEUS_ASN_LABEL})
+    `;
+    const queryResponse = await this.prometheusClient.query(query);
+    const result: PrometheusQueryResult = {};
+    for (const entry of queryResponse.result) {
+      const serializedKey = JSON.stringify(entry.metric, Object.keys(entry.metric).sort());
```

Comment:
It's hard to understand the new logic. I found the original code more readable.

What are you doing with this Stringfy trick? Are you sorting the fields?

Also, what's the meaning of this key?
You are introducing some new complicated concepts here.
I don't actually get why you need a key at all.

Why not just return the prometheus response? This high level logic doesn't seem to belong here, but in the caller instead.

Reply:
Ah, I think you are extracting the location and asn?

Perhaps you make is more explict: `const serializedKey = [metric[PROMETHEUS_ASN_LABEL], metric[PROMETHEUS_COUNTRY_LABEL]]`

Reply:
Still, you don't need a map, just a list. The results are already grouped by the key.